src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -0,0 +1,344 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gcs_actor_scheduler.h"
+#include <ray/common/ray_config.h>
+#include <ray/gcs/redis_gcs_client.h>
+#include <ray/protobuf/node_manager.pb.h>
+#include <ray/util/asio_util.h>
+#include "gcs_actor_manager.h"
+
+namespace ray {
+namespace gcs {
+
+void GcsActorScheduler::GcsLeasedWorker::CreateActor(
+    const TaskSpecification &actor_creation_task, const std::function<void()> &on_done) {
+  RAY_CHECK(on_done);
+  assigned_actor_id_ = actor_creation_task.ActorCreationId();
+
+  std::unique_ptr<rpc::PushTaskRequest> request(new rpc::PushTaskRequest());
+  request->set_intended_worker_id(address_.worker_id());
+  request->mutable_task_spec()->CopyFrom(actor_creation_task.GetMessage());
+
+  google::protobuf::RepeatedPtrField<rpc::ResourceMapEntry> resources;
+  for (auto resource : resources_) {
+    resources.Add(std::move(resource));
+  }
+  request->mutable_resource_mapping()->CopyFrom(resources);
+
+  auto retry_creating_actor = [this, actor_creation_task, on_done] {
+    // Reset the client so that we can reconnect when retry.
+    ResetClient();
+    // The purpose of using the weak_ptr here is to make the worker's life cycle easier to
+    // manage.
+    // Capturing the shared_ptr in lambda is another choice, but it is a little more
+    // complexity because it need add a flag to the worker, if the worker is dead we
+    // should mark the flag to dead, and in the callback of io event we should do like
+    // this:
+    //    if (!shared_worker->is_dead) { do_something(); }
+    std::weak_ptr<GcsLeasedWorker> weak_this(shared_from_this());
+    execute_after(io_context_,
+                  [weak_this, actor_creation_task, on_done] {
+                    if (auto worker = weak_this.lock()) {
+                      worker->CreateActor(actor_creation_task, on_done);
+                    }
+                  },
+                  RayConfig::instance().gcs_create_actor_retry_interval_ms());
+  };
+
+  std::weak_ptr<GcsLeasedWorker> weak_this(shared_from_this());
```

Comment:
> Why do we need to use a weak_ptr here?

Actually, weak_ptr is very useful sometimes, but since both you and @stephanie-wang confused about this, I'll replace it with shared_ptr, and then check if assigned_actor_id is nil to clean up lambda in the event loop at the bottom layer.